### PR TITLE
ALTAPPS-530: iOS show learn next topic on home screen

### PR DESCRIPTION
**Pull request by vladkash**
**YouTrack Issues**:
[#ALTAPPS-530](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-530)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] All checks have been passed;
- [x] Changes have been checked locally.

**Description**
- Added `isLearnNext` flag to `TopicToDiscoverNextButtonView` and badge with arrow icon depending on this flag
- Added `TopicToDiscoverNextCardView` for displaying button with text on home screen
- Fixed condition in shared for updating next topic to learn - **completion** topic instead **repetition**

### CODE REVIEW

**Comment by ivan-magda** [comment]
@vladkash Android PR #335 merged into `develop`, so please update this PR

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Home/Views/HomeView.swift [hunk 1]
```diff
@@ -87,6 +87,11 @@ struct HomeView: View {
                         )
                     }
 
+                    TopicToDiscoverNextCardView(
+                        state: viewModel.topicsToDiscoverNextStateSk,
+                        delegate: viewModel
+                    )
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/Home/Views/HomeView.swift:93]: Here we should add an extra `16` to the top padding.

See screenshot from Figma (this card is like the start of another section -> bigger padding to show this visually)

Screenshot 2023-02-07 at 3 25 52 PM

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardDelegate.swift [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import Foundation
+
+protocol TopicToDiscoverNextCardDelegate: AnyObject {
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardDelegate.swift:3]: Just for your information, we are following Apple's naming for Delegate's
https://vyahhi.myjetbrains.com/youtrack/articles/ALTAPPS-A-30/Swift-Style-Guide#delegaty

Comment (ivan-magda) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardDelegate.swift:3]: and we are using `do` prefix when we call ViewModel directly `viewModel.doLoadContent()`

Comment (ivan-magda) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardDelegate.swift:3]: Also, in the SwiftUI world delegate pattern is not common, but the delegate pattern is commonly used in the UIKit. So when we are working with SwiftUI -> callbacks are preferred, when working with UIKit -> delegate pattern is preferred.

Comment (ivan-magda) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardDelegate.swift:3]: @vladkash for now it's OK (**not necessary to update the code**), but in the future please use info from the comments above.

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardView.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    struct Appearance {
+        let skeletonHeight: CGFloat = 100
+    }
+}
+
+struct TopicToDiscoverNextCardView: View {
+    private(set) var appearance = Appearance()
+
+    let state: TopicsToDiscoverNextFeatureStateKs
+    weak var delegate: (TopicToDiscoverNextCardDelegate)?
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardView.swift:14]:
```suggestion
    weak var delegate: TopicToDiscoverNextCardDelegate?
```

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardView.swift [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    var body: some View {
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardView.swift:16]: Please see how state handling/rendering is implemented in the [TrackTopicsToDiscoverNextBlockView.body](https://github.com/hyperskill/mobile-app/blob/develop/iosHyperskillApp/iosHyperskillApp/Sources/Modules/Track/Views/DiscoverNext/TrackTopicsToDiscoverNextBlockView.swift#L28-L48)

I think something similar would be better

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardView.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                        }
+                    )
+                }
+            }
+        }
+    }
+
+    // MARK: Private API
+
+    private var backgroundSurfaceButtonStyle: OutlineButtonStyle {
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/TopicToDiscoverNextCard/TopicToDiscoverNextCardView.swift:53]: you can move this button style to the `Appearance`

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Track/Views/DiscoverNext/TopicToDiscoverNextButtonView.swift [hunk 6]
```diff
@@ -49,24 +56,61 @@ struct TopicToDiscoverNextButtonView: View {
...
-                    }
+                    buildCompletenessBar(
+                        completenessPercentage:
+                            topic.progress?.completenessPercentage ?? 0
+                    )
                 )
             }
         )
         .buttonStyle(appearance.buttonStyle)
+        .overlay(learnNextBadge)
```
Comment (ivan-magda) [inline iosHyperskillApp/iosHyperskillApp/Sources/Modules/Track/Views/DiscoverNext/TopicToDiscoverNextButtonView.swift:79]: The "Learn next" badge does not animate on tap

Comment (ivan-magda) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/Track/Views/DiscoverNext/TopicToDiscoverNextButtonView.swift:79]: Right now we can not resolve this, please add `TODO`

File: shared/src/commonMain/kotlin/org/hyperskill/app/home/injection/HomeComponentImpl.kt [hunk 7]
```diff
@@ -33,5 +35,5 @@ class HomeComponentImpl(private val appGraph: AppGraph) : HomeComponent {
             gamificationToolbarComponent.gamificationToolbarActionDispatcher,
             topicsToDiscoverNextComponent.topicsToDiscoverNextReducer,
             topicsToDiscoverNextComponent.topicsToDiscoverNextActionDispatcher
-        )
+        ).transformState(HomeFeatureViewStateMapper::map)
```
Comment (ivan-magda) [inline shared/src/commonMain/kotlin/org/hyperskill/app/home/injection/HomeComponentImpl.kt:38]: I think this is an artifact from when merging the remote origin branch, please revert this change.

This line of code already exists in the `HomeFeatureBuilder.build`

File: iosHyperskillApp/iosHyperskillApp/Sources/Modules/Track/Views/DiscoverNext/TopicToDiscoverNextButtonView.swift (hunk 6 above)
Comment (vladkash) [reply iosHyperskillApp/iosHyperskillApp/Sources/Modules/Track/Views/DiscoverNext/TopicToDiscoverNextButtonView.swift:79]: I tried to use the overlay in button label in order to apply animation to the badge but button label crops the badge so I decided to implement it in this way for now

**Review by ivan-magda — APPROVED** [review]
